FF129 SubtleCrypto - support for Ed25519 (#34989)
mdn/content: files/en-us/web/api/subtlecrypto/index.md

Reviewer: Another side note: I always felt we are missing the name context on this page at the top: https://w3c.github.io/webcrypto/#subtlecrypto-interface

>The [SubtleCrypto](https://w3c.github.io/webcrypto/#dfn-SubtleCrypto) interface is named "SubtleCrypto" to reflect the fact that many of these algorithms have subtle usage requirements in order to provide the required algorithmic security guarantees.
>
>For example, the direct use of an unauthenticated encryption scheme, such as [AES in counter mode](https://w3c.github.io/webcrypto/#aes-ctr), gives potential attackers the ability to manipulate bits in the output by manipulating bits in the input, compromising the integrity of the message. However, AES-CTR can be used securely in combination with other cryptographic primitives, such as message authentication codes, to ensure the integrity of the protected message, but only when the message authentication code is constructed over the encrypted message and IV.
>
>Developers making use of the SubtleCrypto interface are expected to be aware of the security concerns associated with both the design and implementation of the various algorithms provided. The raw algorithms are provided in order to allow developers maximum flexibility in implementing a variety of protocols and applications, each of which may represent the composition and security parameters in a unique manner that necessitate the use of the raw algorithms.

Even paraphrasing the first paragraph might be nice, what do you reckon?

Author: @bsmth This is captured a bit in the following warning, but I also think it is good to highlight. Done.

#### Storing keys

`CryptoKey` is a {{glossary("serializable object")}}, which allows keys to be stored and retrieved using standard web storage APIs.

> The specification expects that most developers will use the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API), sotring `CryptoKey` objects against some key string identifier that is meaningful to the application, along with any other metadata it finds useful.

Reviewer: ```suggestion
The specification expects that most developers will use the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API), sorting `CryptoKey` objects against some key string identifier that is meaningful to the application, along with any other metadata it finds useful, if needed.
This allows the storage and retrieval of key material without exposing it to the application or the JavaScript environment.
```

I wonder was there a reason this was left out originally? This wording comes from the spec.

Author: Yes, it means storing.

```suggestion
The specification expects that most developers will use the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API), storing `CryptoKey` objects against some key string identifier that is meaningful to the application, along with any other metadata it finds useful.
```
1. ", if needed" is redundant.

> `CryptoKey` is a {{glossary("serializable object")}}, which allows keys to be stored and retrieved using standard web storage APIs.
> The specification expects that most developers will use the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API), storing `CryptoKey` objects against some key string identifier that is meaningful to the application, along with any other metadata it finds useful.

Author: @bsmth , following on from https://github.com/mdn/content/pull/34989/files#r1698336482

>> This allows the storage and retrieval of key material without exposing it to the application or the JavaScript environment.

> I wonder was there a reason this was left out originally? This wording comes from the spec.  
   
Ah, I chose not to include it because to me it didn't make sense/wasn't clear how this could work.

But now it does - when you understand that it is `CryptoKey` hides the key material.

```suggestion
`CryptoKey` is a {{glossary("serializable object")}}, which allows keys to be stored and retrieved using standard web storage APIs.

The specification expects that most developers will use the [IndexedDB API](/en-US/docs/Web/API/IndexedDB_API), storing `CryptoKey` objects against some key string identifier that is meaningful to the application, along with any other metadata it finds useful.
This allows the storage and retrieval of the `CryptoKey` without having to expose its underlying key material to the application or the JavaScript environment.
```

### Supported algorithms

> 

Reviewer: In terms of choosing an algo, should we mention that it varies by browser or engine **which** algorithm is implemented / supported, so check BCD? There may be others not listed below, but these are the recommended ones (per standards?) and proceed with caution?

<hr>

Side note about caution, there's a section about deciding what to use and when for developers: https://www.w3.org/TR/WebCryptoAPI/#security-developers

>some of which have known weaknesses when used inappropriately. Application developers must take care and review appropriate and current cryptographic literature, to understand and mitigate such issues. 

I know we have a warning on each of the top-level pages here, the points from the spec might also be good. Here's the summary for developers in a nutshell:

* choose an algo based on it's known weaknesses and don't use it for things it's known to be bad at
* have a holistic view of your app's security, because these algos can be undermined by other weaknesses in your system/application
* There's no guarantees that an implementation __will not persist__ key material
* Recovery may be possible if not persisted (user agents are not required to zeroize key material)
* Using a key across origins ([`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage)) and other origins not being able to revoke keys, so handling cross-origin things can be complex
* If a user clears the storage associated with an origin, they may destroy keys

Reviewer 2: > In terms of choosing an algo, should we mention that it varies by browser or engine which algorithm is implemented / supported, so check BCD?

I'm -1 on doing this in general. People _always_ have to check BCD. 

> There may be others not listed below, but these are the recommended ones (per standards?) and proceed with caution?

Which others? This should be an exhaustive list of algorithms supported in the Web Crypto API (I mean sure you could use RC4 for encryption, but not in the Web Crypto API).

Author: As Will says, you don't need to advise BCD check, not just because they should always check, but because the unsupported algorithms fail the call, and a developer should be coding for that. To me that makes it redundant.

W.r.t. the security information I think the notes up the top of each page catch the most important comment - which is you need to develop some expertise and this is beyond what MDN can give you.
Some of the points probably make sense to capture, but they are "out of scope", and I don't have the expertise to do anything but parrot them.

So do not plan to do anything in response to this.

TLDR - a few opinions on the security points:

> "choose an algo based on it's known weaknesses and don't use it for things it's known to be bad at

Would be good to have a comparative guide about what algorithms to choose.

> have a holistic view of your app's security, because these algos can be undermined by other weaknesses in your system/application

Common sense, but it's a top level hand-wavy crypto overview page thing, not a "use this API thing".

> There's no guarantees that an implementation will not persist key material

True, but unhelpful.

A developer basically either has to assume that the material is safe, or not use the API. A security expert may know how to assess the importance vs risk, but if so, they will still have no idea whether or not they can use the API - all they can do is make big decisions like "I won't put the root private key for a certificate authority on this device"

> Recovery may be possible if not persisted (user agents are not required to zeroize key material)

That's a bit obvious. The keys are javascript values in volatile memory until you save them. 

Not sure about the others.

Reviewer 2: Regarding guidance for the web crypto API, a bit of history: https://github.com/mdn/sprints/issues/672 (also https://github.com/mdn/sprints/issues/664).

Author: Looks like you came to similar conclusions that I did "good idea, tough to write, needs expertise to write well".

      <td></td>
      <td></td>
      <td>✓</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/digest#supported_algorithms">SHA-512</a></th>
      <td></td>
      <td></td>
      <td>✓</td>
      <td></td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/deriveKey#ecdh">ECDH</a></th>
      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/deriveKey#hkdf">HKDF</a></th>
      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/deriveKey#pbkdf2">PBKDF2</a></th>
      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
      <td></td>
      <td></td>
    </tr>
    <tr>
      <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/deriveKey#x25519">X25519</a></th>
      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
      <td>✓</td>

>       <td>✓</td>

Reviewer 2: I think this table update is a bit wrong. It's hard to tell because the spec isn't updated, but:

- it looks like Ed25519 is a signing algo, so should have the same usages as ECDSA, in particular NOT wrapKey/unwrapKey, which should just be for encryption algos.
- it looks like X25519 is a key agreement algo, so it should have the same usages as ECDH, again NOT wrapKey/unwrapKey.
- ECDSA and ECDH should have generateKey/importKey/exportKey (basically AFAICT all the ones that aren't pure hash functions should have these),.
- HKDF and PBKDF2 have importKey but NOT generateKey or exportKey, which is unfortunate as you have to split that column and break the page layout :(.

Author: @wbamberg Thank you. I must have misinterpreted that they were supported in [this comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1804788#c27).

That pointed me to [the tests](https://wpt.fyi/results/WebCryptoAPI/wrapKey_unwrapKey/wrapKey_unwrapKey.https.any.html?label=experimental&label=master&aligned)  - which seemed to indicate support, but in retrospect show that you can wrap a key created with the algorithm, not that it is used by the method.

I have updated as you suggested. Note that splitting out importKey into its own column caused overflow. I have reduced this by removing the `()` from the function names - I don't see this as confusing or harmful.

Reviewer 2: >  I have reduced this by removing the () from the function names - I don't see this as confusing or harmful.

+1

    <tr>
      <th scope="row"></th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/sign">sign</a><br /><a href="/en-US/docs/Web/API/SubtleCrypto/verify">verify</a>
      </th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/encrypt">encrypt</a><br /><a href="/en-US/docs/Web/API/SubtleCrypto/decrypt">decrypt</a>
      </th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/digest">digest</a>
      </th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/deriveBits">deriveBits</a><br /><a href="/en-US/docs/Web/API/SubtleCrypto/deriveKey">deriveKey</a>
      </th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/wrapKey">wrapKey</a><br /><a href="/en-US/docs/Web/API/SubtleCrypto/unwrapKey">unwrapKey</a>
      </th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/generateKey">generateKey</a><br /><a href="/en-US/docs/Web/API/SubtleCrypto/exportKey">exportKey</a>
      </th>
      <th scope="col">
        <a href="/en-US/docs/Web/API/SubtleCrypto/importKey">importKey</a>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>

>       <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/sign#rsassa-pkcs1-v1_5">RSASSA-PKCS1-v1_5></a></th>

Reviewer 2: ```suggestion
      <th scope="row"><a href="/en-US/docs/Web/API/SubtleCrypto/sign#rsassa-pkcs1-v1_5">RSASSA-PKCS1-v1_5</a></th>
```

      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
    </tr>
    <tr>
      <th scope="row">Ed25519</th>
      <td>✓</td>
      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
      <td>✓</td>
    </tr>
    <tr>
      <th scope="row">X25519</th>
      <td></td>
      <td></td>
      <td></td>
      <td>✓</td>
      <td>✓</td>
      <td>✓</td>

>     </tr>

Author: This is update to table which originates here https://www.w3.org/TR/WebCryptoAPI/#algorithm-overview + addition of the two new algorithms. (not yet shown in this spec)
